pkg/container/versioned/value_test.go

```diff
@@ -0,0 +1,388 @@
+
+	assert.Empty(t, value1.At(handle))
+	assert.Empty(t, value2.At(handle))
+
+	v = value1.At(handle3)
+	assert.Equal(t, []uint32{100, 150, 200}, v)
+
+	v = value2.At(handle3)
+	assert.Equal(t, []uint32{110, 150, 190}, v)
+
+	v = value1.At(handle2)
+	assert.Equal(t, []uint32{100, 200}, v)
+
+	v = value2.At(handle2)
+	assert.Equal(t, []uint32{110, 190}, v)
+
+	// handle closers can be called in any order
+	assert.Nil(t, handle2.Close())
+	assert.Nil(t, handle3.Close())
+	assert.Nil(t, handle1.Close())
+	assert.Nil(t, handle4.Close())
+
+	// old value1 have been cleaned off
+	cleaner.waitUntilCleaned(handle3.version)
+	assert.Nil(t, value1.head.Load())
+
+	// but value2 remains, as it was not removed
+	assert.NotNil(t, value2.head.Load())
+}
+
+func TestPairSlice(t *testing.T) {
+	cv := Coordinator{}
+	s := make(VersionedSlice[int], 0, 10)
+
+	// 1st value '2000' at version 1
+	tx := cv.PrepareNextVersion()
+	s = s.Append(2000, tx)
+	assert.Len(t, s, 1)
+	assert.Equal(t, version(1), s[0].version)
+	assert.Equal(t, 2000, s[0].value)
+
+	tx.Commit()
+	tx = cv.PrepareNextVersion()
+
+	// 2nd value '1000' at version 2
+	s = s.Append(1000, tx)
+	assert.Len(t, s, 2)
+	assert.Equal(t, version(1), s[0].version)
+	assert.Equal(t, 2000, s[0].value)
+	assert.Equal(t, version(2), s[1].version)
+	assert.Equal(t, 1000, s[1].value)
+
+	tx.Commit()
+	tx = cv.PrepareNextVersion()
+
+	// 3rd value '3000' at version 3
+	s = s.Append(3000, tx)
+	assert.Len(t, s, 3)
+	assert.Equal(t, version(1), s[0].version)
+	assert.Equal(t, 2000, s[0].value)
+	assert.Equal(t, version(2), s[1].version)
+	assert.Equal(t, 1000, s[1].value)
+	assert.Equal(t, version(3), s[2].version)
+	assert.Equal(t, 3000, s[2].value)
+
+	// get all values before version 3; excluding the last value
+	var values []int
+	n := s.ForEachBefore(KeepVersion(3), func(i int) {
+		values = append(values, i)
+	})
+	assert.Equal(t, []int{2000, 1000}, values)
+	s = s[n:]
+
+	// get all values upto and including maxVersion; get the lone value
+	values = nil
+	n = s.ForEachBefore(KeepVersion(invalidVersion), func(i int) {
+		values = append(values, i)
+	})
+	assert.Equal(t, []int{3000}, values)
+	s = s[n:]
+	assert.Len(t, s, 0)
+}
+
+func TestVersionedChaos(t *testing.T) {
+	const nValues = 10
+	var values [nValues]Value[[]uint32]
+
+	cleaner := newTestCleaner()
+
+	cv := Coordinator{
+		Cleaner: func(keepVersion KeepVersion) {
+			for i := 0; i < nValues; i++ {
+				cleaner.cleanValue(keepVersion, &values[i])
+			}
+		},
+	}
+
+	// Initially empty
+	handle := cv.GetVersionHandle()
+	for i := 0; i < nValues; i++ {
+		assert.Empty(t, values[i].At(handle))
+	}
+	assert.Nil(t, handle.Close())
+	assert.NotNil(t, handle.Close())
+
+	var mutex lock.Mutex
+	var wg sync.WaitGroup
+	for j := 0; j < 1000; j++ {
+		for k := 0; k < 100; k++ {
+			go func() {
+				version := cv.GetVersionHandle()
+				time.Sleep(time.Duration(rand.IntN(100)) * time.Millisecond)
+				fmt.Printf("Closing version %s\n", version)
```

Comment:
nit: if you use `t.Logf` in the tests, the logging only shows up when the test fails.

Reply:
Changed, thanks!